README.md
# VSCode-Pomodoro

Better Manage your time and be more effecient with your development work.

For more information about the Pomodoro Technique: [Visit this link](https://cirillocompany.de/pages/pomodoro-technique)


## Installation

Press F1 in VSCode, type 
```
ext install vscode-pomodoro
```

## Usage
![Usage](images/usage.gif)
### Add tasks
First you'll need to add tasks to the extenstion to time-manage it for you.
 
> ``` Pomodoro: Add New Task ```

### Start The Pomodoro
Simply by clicking on the Green Button next to the timer in the status bar, the pomodoro cycles start and you're good to go... 

The tasks added to the extension are persisted throughout vscode windows.
The counter in the status bar shows the number of completed tasks over the number of tasks.

you many, sometime, start with a clean slate and reset the counter of the completed tasks.
you can very easily do that with the following command:
``` Pomodoro: Clear Completed Tasks```
Dont't worry.. Tasks that hasn't been completed will remain intact.

## Extension Settings

This extension contributes the following settings:

* `"pomodoro.task_duarion"`: The default duration of a Pomodoro Task in minutes
* `"pomodoro.break_duarion"`: The default duration of a Break in minutes
* `"pomodoro.long_break_duarion"`: The default duration of a long Break in minutes
* `"pomodoro.counter_to_long_break"`: number of regular breaks before a long break
* `"pomodoro.tasks_file"`: full path to the tasks.json file

## Known Issues

Please Report any issue to the Git repository.

The extension was only tested on macOS, some issues may arise on other Operating systems. 

Please feel free to file new issues and submit pull requests to resolve any issue you may have.


## Release Notes

please refer to the change log CHANGELOG.md

**Enjoy!**

## License

MIT © Yahya Gilany


CHANGELOG.md
# Change Log
All notable changes to the "vscode-pomodoro" extension will be documented in this file.

Check [Keep a Changelog](http://keepachangelog.com/) for recommendations on how to structure this file.

## [Unreleased]
- added a pause functionality to the pomodoro
- added the ability to finish a task early

## 1.0.1 - 1.0.2 - 1.0.3
- documentation changes

## 1.0.0
- Initial release

src/task.ts
import { TimeUnits, Timer } from './timer';
import { getConfig } from './config';

export class Task {
  public name: string;
  public startTime: string;
  public pauseTime: string;
  public isCompleted: boolean;

  constructor(_name: string, _startTime: string, _isCompleted: boolean = false) {
    this.name = _name;
    this.startTime = _startTime;
    this.isCompleted = _isCompleted;
  }

  public CompleteTask() {
    this.isCompleted = true;
  }

  public updateTaskName(newName: string) {
    this.name = newName;
  }

  public PauseTask() : void {
    this.pauseTime = `${new Date()}`;
  }

  public startTask(next: Function) : Timer {
    let duration = getConfig().task_duration;    
    if (this.startTime !== null && this.startTime !== undefined){ // if the task already started      
      let difference: number;

      if(this.pauseTime !== null && this.pauseTime !== undefined) { // task was paused        
        difference = new Date(this.pauseTime).getTime() - new Date(this.startTime).getTime();

        let newStartTime = new Date();
        newStartTime.setTime(newStartTime.getTime() - (new Date(this.pauseTime).getTime() - new Date(this.startTime).getTime()));
        
        this.startTime = `${newStartTime}`;
      } else {        
        difference = new Date().getTime() - new Date(this.startTime).getTime()
      }
      duration -= difference      
    } else {
      this.startTime = `${new Date()}`;      
    }
    this.pauseTime = undefined;
    let _timer = new Timer(duration, TimeUnits.Milliseconds);
    _timer.start(next);
    
    return _timer;
  }
}

src/timer.ts
const MINUTES_IN_HOUR: number = 60;
const SECONDS_IN_MINUTE: number = 60;
const MILLISECONDS_IN_SECOND : number = 1000;

import { StatusBar } from './ui';

export enum TimeUnits {
  Milliseconds = 1,
  Second = MILLISECONDS_IN_SECOND,
  Minute = SECONDS_IN_MINUTE * MILLISECONDS_IN_SECOND,
  Hour = MINUTES_IN_HOUR * SECONDS_IN_MINUTE * MILLISECONDS_IN_SECOND
};

export class Timer {
  public countdownMilliseconds: number;
  private isRunning: boolean;
  private _timer: NodeJS.Timer;


  constructor(countdown: number = 0, unit:TimeUnits) {
    this.countdownMilliseconds = countdown * unit;
  }

  public start (next: Function) {
    const statusBars = StatusBar.getInstance();
    if (!this.isRunning){
      this.isRunning = true;
      this._timer = setInterval(()=> {
        this.tick();
        statusBars.updateTimerBar(this.countdownMilliseconds);
        
        if(this.countdownMilliseconds <= 0) {
          this.stop()
          next();
        }
      }, TimeUnits.Second);
    }
  }

  public stop() {
    if(this.isRunning) {
      this.isRunning = false;
      clearInterval(this._timer);
    }
  }

  public reset() {
    if(this.isRunning) {
      this.stop();
      this.countdownMilliseconds = 0;
    }
  }

  public tick() {    
    this.countdownMilliseconds -= TimeUnits.Second;
  }
}

src/config.ts
import { workspace } from 'vscode';
import { TimeUnits } from './timer';

import path = require('path');

const DEFAULT_TASK_DURATION = 25 * TimeUnits.Minute;
const DEFAULT_BREAK_DURATION = 5 * TimeUnits.Minute;
const DEFAULT_LONG_BREAK_DURATION = 15 * TimeUnits.Minute;

const COUNTER_TO_LONG_BREAK = 3;

export interface Config {
    task_duration: number,
    break_duration: number,
    long_break_duration: number,
    counter_to_long_break: number, 
    tasks_file: string
}

export function getConfig(): Config {
    let configuration;
    try {
        configuration = workspace.getConfiguration('pomodoro');
    } catch (error) {}    
    
    return {
      task_duration: configuration.task_durarion * TimeUnits.Minute || DEFAULT_TASK_DURATION,
      break_duration: configuration.break_duration * TimeUnits.Minute || DEFAULT_BREAK_DURATION,
      long_break_duration: configuration.long_break_duration * TimeUnits.Minute || DEFAULT_LONG_BREAK_DURATION,
      counter_to_long_break: configuration.counter_to_long_break || COUNTER_TO_LONG_BREAK, 
      tasks_file: configuration.task_file || path.join(__dirname, `../../tasks.json`)
    } as Config;
}

src/storage.ts
import fs = require('fs');
import { Task } from './task';
import { Pomodoro } from './pomodoro';

export class TaskStorage {
  private filename: string;
  private _pomodoro: Pomodoro;

  constructor(_filename: string) {
    this.filename = _filename;
  }

  public save(): void {
    if (this._pomodoro === undefined) {
      this._pomodoro = Pomodoro.getInstance();
    }
    fs.writeFileSync(this.filename, JSON.stringify(this._pomodoro.tasks, null, "\t"));
  }

  public load(): void {
    let tasks: Task[] = [];

    if (this._pomodoro === undefined) {
      this._pomodoro = Pomodoro.getInstance();
    }

    if (!fs.existsSync(this.filename)) {
      this._pomodoro.tasks = tasks
    }

    try {
      let items = JSON.parse(fs.readFileSync(this.filename).toString());

      tasks = items.map(t => {
        return new Task(t.name, t.startTime, t.isCompleted)
      });
      
      this._pomodoro.tasks = tasks;

    } catch (error) {
      console.log(error);
    }
  }

}

src/pomodoro.ts
import * as vscode from 'vscode';
import { Task } from './task';
import { TimeUnits, Timer } from './timer';
import { getConfig } from './config';
import { YesNoPrompt, InputPrompt, StatusBar } from './ui';
import { TaskStorage } from './storage';


export class Pomodoro {
	private static _instance: Pomodoro;

	private _statusBars: StatusBar = StatusBar.getInstance();

	private _storage: TaskStorage;

	public tasks: Task[];
	public completedTasksCounter: number;
	public currentTaskIndex: number;

	private breakCounter: number;

	private _timer: Timer;

	private constructor() {
		this.tasks = [] as Task[];
		this.completedTasksCounter = 0;
		this.breakCounter = 0;
		this._storage = new TaskStorage(getConfig().tasks_file);
	}

	public static getInstance(): Pomodoro {		
		if (Pomodoro._instance === null || Pomodoro._instance === undefined) {
			Pomodoro._instance =  new Pomodoro();
		}
		return Pomodoro._instance;
	}

	public preload() {
		const pomodoro = Pomodoro.getInstance();
		pomodoro._storage.load();

		for (let taskIndex in pomodoro.tasks) {			
			if (pomodoro.tasks[taskIndex].startTime === null) {				
				break;
			} else {				
				if (pomodoro.tasks[taskIndex].isCompleted) {
					pomodoro.completedTasksCounter ++;
				} else {
					pomodoro.currentTaskIndex = parseInt(taskIndex);
				}
			}
		}
		
		if (pomodoro.currentTaskIndex !== undefined && pomodoro.tasks[pomodoro.currentTaskIndex].startTime !== null) {
			pomodoro.run();
		}
	}

	public async addTask() {
		const pomodoro = Pomodoro.getInstance();
		const newTask: string = await InputPrompt(`Add a new task to the Pomodoro`, `task name`);

		pomodoro.tasks.push(new Task(newTask, null));
		pomodoro._storage.save();

		pomodoro._statusBars.updateTasksCounter(pomodoro.completedTasksCounter, pomodoro.tasks.length)
	}

	public run() {
		const pomodoro = Pomodoro.getInstance();
		pomodoro.pickTask();
		if (pomodoro.currentTaskIndex < pomodoro.tasks.length) {
			pomodoro._statusBars.updateCurrentTask(`Focus: `+pomodoro.tasks[pomodoro.currentTaskIndex].name)
		
			pomodoro._timer = pomodoro.tasks[pomodoro.currentTaskIndex].startTask(pomodoro.askAboutTaskCompletion);
			pomodoro._statusBars.updateStartBar();
			pomodoro._storage.save();	
		} else {
			return;
		}		
	}

	public pause() {
		const pomodoro = Pomodoro.getInstance();
		pomodoro.tasks[pomodoro.currentTaskIndex].PauseTask();
		pomodoro._timer.stop();
		pomodoro._statusBars.updateStartBar();
		pomodoro._storage.save();		
	}

	public stopTask() {
		const pomodoro = Pomodoro.getInstance();
		pomodoro._timer.reset();

		pomodoro.finishTask();
		pomodoro._storage.save();
		pomodoro.takeBreak();
	}

	private pickTask(): void {
		const pomodoro = Pomodoro.getInstance();
		if (pomodoro.tasks.length > 0) {
			if (pomodoro.currentTaskIndex === undefined) {
				pomodoro.currentTaskIndex = 0;
			} else {
				if (pomodoro.tasks[pomodoro.currentTaskIndex].isCompleted) {
					pomodoro.currentTaskIndex += 1;
				}
			}
		}
	}

	private async askAboutTaskCompletion() {
		const pomodoro = Pomodoro.getInstance();
		const response: boolean = await YesNoPrompt(`Did you finish the task?`);
		if(response) {
			pomodoro.finishTask();
		}
		pomodoro.takeBreak();
	}

	private takeBreak(): void {
		const pomodoro = Pomodoro.getInstance();		
		if (pomodoro.breakCounter < getConfig().counter_to_long_break) {
			pomodoro._timer = new Timer(getConfig().break_duration, TimeUnits.Milliseconds);
			pomodoro.breakCounter++;
		} else {
			pomodoro._timer = new Timer(getConfig().long_break_duration, TimeUnits.Milliseconds);
			pomodoro.breakCounter = 0;
		}
		pomodoro._statusBars.updateCurrentTask(`Break 💃🏻`);
		pomodoro._timer.start(pomodoro.run);
	}

	public clearCompleted(): void {
		const pomodoro = Pomodoro.getInstance();
		pomodoro.tasks = pomodoro.tasks.filter(function (task) {
			return !task.isCompleted;
		});

		pomodoro.completedTasksCounter = 0;
		pomodoro.currentTaskIndex = undefined;
		pomodoro._storage.save();
		pomodoro._statusBars.updateTasksCounter(pomodoro.completedTasksCounter, pomodoro.tasks.length)
	}

	private finishTask(): void {
		const pomodoro = Pomodoro.getInstance();
		pomodoro.tasks[pomodoro.currentTaskIndex].CompleteTask();
		pomodoro.completedTasksCounter ++;
		pomodoro._statusBars.updateTasksCounter(pomodoro.completedTasksCounter, pomodoro.tasks.length)
		pomodoro._storage.save();
	}
}

src/extension.ts
import * as vscode from 'vscode';
import { Pomodoro } from './pomodoro';
import { StatusBar } from './ui';

export function activate(context: vscode.ExtensionContext) {
	const pomodoro = Pomodoro.getInstance();
	const statusBars = StatusBar.getInstance();

	pomodoro.preload();
	
	statusBars.updateTasksCounter(pomodoro.completedTasksCounter, pomodoro.tasks.length);

	vscode.commands.registerCommand(`pomodoro.addTask`, () => pomodoro.addTask());
	vscode.commands.registerCommand(`pomodoro.run`, () => pomodoro.run());	
	vscode.commands.registerCommand(`pomodoro.clear`, () => pomodoro.clearCompleted());	
	vscode.commands.registerCommand(`pomodoro.pause`, () => pomodoro.pause());
	vscode.commands.registerCommand(`pomodoro.finishTask`, () => pomodoro.stopTask());
}

export function deactivate() {}

src/ui.ts
import * as vscode from 'vscode';

export class StatusBar {
  private static _instance: StatusBar;

  private static timerStatusBar: vscode.StatusBarItem;
  private static tasksCounterStatusBar: vscode.StatusBarItem;
  private static taskStatusBar: vscode.StatusBarItem;
  private static startStatusBar: vscode.StatusBarItem;

  private constructor() {
    StatusBar.timerStatusBar = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 3);
    StatusBar.tasksCounterStatusBar = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 2);
    StatusBar.taskStatusBar = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 1);
    StatusBar.startStatusBar = vscode.window.createStatusBarItem(vscode.StatusBarAlignment.Right, 4);

    StatusBar.timerStatusBar.text = `$(watch) 00:00`;

    StatusBar.taskStatusBar.text = `nothing is currently running`;
    StatusBar.tasksCounterStatusBar.text = `(0/0)`;

    StatusBar.startStatusBar.text = `$(triangle-right)`;
    StatusBar.startStatusBar.color = 'lightgreen';
    StatusBar.startStatusBar.tooltip = 'start';
    StatusBar.startStatusBar.command = 'pomodoro.run';

    StatusBar.timerStatusBar.show();
    StatusBar.tasksCounterStatusBar.show();
    StatusBar.taskStatusBar.show();
    StatusBar.startStatusBar.show();

  }

  public static getInstance(): StatusBar {
    if (StatusBar._instance === undefined ){
      StatusBar._instance = new StatusBar();
    }
    return StatusBar._instance;
  }

  public updateTasksCounter(completedTasks: number = 0, totalTasks: number) {
    StatusBar.tasksCounterStatusBar.text = `(${completedTasks}/${totalTasks})`
  }

  public updateTimerBar(milliseconds) {
    StatusBar.timerStatusBar.text = `$(watch) ${this.convertMS(milliseconds)}`
  }

  public updateCurrentTask(name: string) {
    StatusBar.taskStatusBar.text = name
  }

  public updateStartBar(){
    if(StatusBar.startStatusBar.text === `$(triangle-right)`) {
      StatusBar.startStatusBar.text = `$(primitive-square)`;
      StatusBar.startStatusBar.color = 'red';
      StatusBar.startStatusBar.tooltip = 'pause';
      StatusBar.startStatusBar.command = 'pomodoro.pause'
    } else {
      StatusBar.startStatusBar.text = `$(triangle-right)`;
      StatusBar.startStatusBar.color = 'lightgreen';
      StatusBar.startStatusBar.tooltip = 'start';
      StatusBar.startStatusBar.command = 'pomodoro.run';
    }
  }

  public convertMS(ms): string {
    function pad(number) {
      return ('00' + number).slice(-2);
    }
    var mins, secs;
    secs = Math.floor(ms / 1000);
    mins = Math.floor(secs / 60);
    secs = secs % 60;

    return  pad(mins) + ':' + pad(secs);
  };

  public dispose() {
    StatusBar.timerStatusBar.dispose();
    StatusBar.tasksCounterStatusBar.dispose();
    StatusBar.taskStatusBar.dispose();
  }
}

export async function YesNoPrompt(prompt: string): Promise<boolean> {
  const optionYes = {
    title: "Yes"
  } as vscode.MessageItem;
  const optionNo = {
    title: "No"
  } as vscode.MessageItem;

  const selection = await vscode.window.showInformationMessage(
    prompt,
    optionYes, optionNo
  );

  if (selection.title === `Yes`) {
    return Promise.resolve(true);
  }else {
    return Promise.resolve(false);
  }
}

export async function InputPrompt(_prompt: string, _placeHolder: string): Promise<string> {
  const response: string = await vscode.window
    .showInputBox({
      prompt: _prompt,
      placeHolder: _placeHolder,
    });

  return response;
}